### PR TITLE
typescript compiler with definition

### PR DESCRIPTION
**Pull request by qballer**
Created a new typescript compiler:
1. Emits d.ts
2. takes a 1/3 of the time 

Please review merge and publish.

### CODE REVIEW

File: components/compilers/typescript/index.js [hunk 1]
```diff
@@ -1,66 +1,127 @@
...
+        return [mappings, distFile, declarationFile];
+    })
+    .filter((a)=> !!a)
+    .reduce((a, b) => a.concat(b), [declarationFile])
+
+    return results
+}
 const _toCompile = (file) => {
-  return compiledFileTypes.indexOf(file.extname.replace('.', '')) > -1 && !file.stem.endsWith('.d');
+    return compiledFileTypes.indexOf(file.extname.replace('.','')) > -1;// && !file.stem.endsWith('.d');
```
Comment (GiladShoham) [inline components/compilers/typescript/index.js]: Are you sure about this change? bit will try to compile d.ts files.. which will makes problems

File: package.json [hunk 2]
```diff
@@ -75,7 +75,7 @@
     "memory-fs": "^0.4.1",
     "mocha": "^5.0.2",
     "mockery": "^2.1.0",
-    "node-sass": "^4.7.2",
+    "node-sass": "^4.12.0",
```
Comment (GiladShoham) [inline package.json]: what and why this change is related?

Comment (qballer) [reply package.json]: node sass 4.7 didn't compiler.

File: components/compilers/typescript/index.js (hunk 1 above)
Comment (qballer) [reply components/compilers/typescript/index.js]: changed

File: components/internal/typescript-base-compiler/typescript-base.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      if (reviseExtension) fileRelative = _getRevisedFileExtension(file.relative);
+  
+      return path.join(distPath, fileRelative);
+  }
+  
+  const _getRevisedFileExtension = (fileName) => {
+      return fileName.replace('.tsx', '.js').replace('.ts', '.js');
+  }
+  
+  function longestStartingSubstring(left, right) {
```
Comment (GiladShoham) [inline components/internal/typescript-base-compiler/typescript-base.js]: There are some nice implementation of this described here:
https://stackoverflow.com/questions/1916218/find-the-longest-common-starting-substring-in-a-set-of-strings
We also implemented it in bit (it should be a component of course)
https://github.com/teambit/bit/blob/master/src/utils/array/shared-start.js
(I don't expect you to change it.. just to let you know about it)
but a comment about what it does will be nice
